internal/auth0/user/resource_permissions.go

```diff
@@ -66,46 +66,56 @@ func NewPermissionsResource() *schema.Resource {
 }
 
 func upsertUserPermissions(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if !data.HasChange("permissions") {
+		return nil
+	}
+
 	api := meta.(*config.Config).GetAPI()
 	mutex := meta.(*config.Config).GetMutex()
 
 	userID := data.Get("user_id").(string)
 
-	if !data.HasChange("permissions") {
-		return nil
-	}
-
 	mutex.Lock(userID)
 	defer mutex.Unlock(userID)
 
 	toAdd, toRemove := value.Difference(data, "permissions")
 
-	var addPermissions []*management.Permission
-	for _, addPermission := range toAdd {
-		permission := addPermission.(map[string]interface{})
-		addPermissions = append(addPermissions, &management.Permission{
+	var rmPermissions []*management.Permission
```

Comment:
The fix is swapping around the removal and additions. We need to always perform removals first.